sys/callout.c

```diff
@@ -142,4 +146,20 @@ void callout_process(systime_t time) {
   ci.last = time;
 }
 
+bool callout_drain(callout_t *handle) {
+  /* A callout may be in active state only in callout_process,
+   * which is called in bottom half (with interrupts disabled),
```

Comment:
I wanted to have interrupts disabled for entire `callout_drain` function, but I don't know how to do it.
* We can't take `SCOPED_SPINLOCK(&ci.lock);` at the beginning because then we will go to sleep with spinlock acquired (another thread will try to take this spinlock and see that its owner `!= NULL` and assert `spinlock.c:24` fails).
* We can't put all the code inside `WITH_INTR_DISABLED { ... }` because it causes compiler error `control reaches end of non-void function`.

As for now, the function behaviour is correct (at least I think so) but I thought that disabling interrupts for entire function would be more clean.

Reply:
You've found an interesting problem. My first (and probably wrong) solution would be to extend `sleepq_wait` to accept optional `spinlock_t *` argument and then unlock it while holding `td_spin` and acquire on return. That would make it similar to `cv_wait`, but handle top-bottom half synchronization. **Do not** implement it, though. That needs more careful thinking that I can give it right now.